### PR TITLE
Implement 46 DRAGONS cards

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Implement 46 DRAGONS cards
  - Scion of Ruin (DRG_019)
  - EVIL Quartermaster (DRG_020)
  - Ritual Chopper (DRG_021)
  - Ramming Speed (DRG_022)
  - Skybarge (DRG_023)
  - Sky Raider (DRG_024)
  - Ancharrr (DRG_025)
  - Deathwing, Mad Aspect (DRG_026)
  - Umbral Skulker (DRG_027)
  - Dragon's Hoard (DRG_028)
  - Praise Galakrond! (DRG_030)
  - Necrium Apothecary (DRG_031)
  - Candle Breath (DRG_033)
  - Bloodsail Flybooter (DRG_035)
  - Waxadred (DRG_036)
  - Strength in Numbers (DRG_051)
  - Veranus (DRG_095)
  - Azure Explorer (DRG_102)
  - Arcane Breath (DRG_106)
  - Violet Spellwing (DRG_107)
  - Sky Claw (DRG_225)
  - Amber Watcher (DRG_226)
  - Bronze Explorer (DRG_229)
  - Lightforged Crusader (DRG_231)
  - Lightforged Zealot (DRG_232)
  - Sand Breath (DRG_233)
  - Dragonrider Talritha (DRG_235)
  - Time Rip (DRG_246)
  - Seal Fate (DRG_247)
  - Awaken! (DRG_249)
  - Clear the Way (DRG_251)
  - Phase Stalker (DRG_252)
  - Dwarven Sharpshooter (DRG_253)
  - Primordial Explorer (DRG_254)
  - Dragonbane (DRG_256)
  - Sanctuary (DRG_258)
  - Whispers of EVIL (DRG_301)
  - Nozdormu the Timeless (DRG_309)
  - Treenforcements (DRG_311)
  - Shrubadier (DRG_312)
  - Emerald Explorer (DRG_313)
  - Aeroponics (DRG_314)
  - Secure the Deck (DRG_317)
  - Breath of Dreams (DRG_318)
  - Ysera, Unleashed (DRG_320)
  - Molten Breath (DRG_500)
- Implement tasks
  - ActivateCapturedDeathrattleTask: This class represents the task for activating captured card's deathrattle.
  - AttackTask: This class represents the task for attacking the target.
  - SummonStackTask: This class represents the task for summoning something from stack.
- Add SelfCondition wrappers
  - HasInvokedTwice(): SelfCondition wrapper for checking the player has invoked twice.
  - HasDeathrattle(): SelfCondition wrapper for checking the entity has deathrattle.
  - HasNoNeutralCardsInDeck(): SelfCondition wrapper for checking the player has no neutral cards in deck.
  - IsMyHeroUndamagedEnemyTurn(): SelfCondition wrapper for checking my hero is undamaged in enemy turn.
- Change the logic of 'Steady Shot' (DS1h_292): Change 'EntityType::ENEMY_HERO' to 'EntityType::TARGET'
  - Add code to process 'PlayReq::REQ_STEADY_SHOT' and 'PlayReq::REQ_MINION_OR_ENEMY_HERO'
- Update functions and methods
  - Add parameter 'skipPrePhase' to function 'Attack()': SelfCondition wrapper for checking the player has no neutral cards in deck.
  - Add parameter 'race' and 'rarity' to 'DiscoverTask': The race and rarity of card to discover.
  - Add variable 'm_capturedCard' to class 'Enchantment'
    - Add getter/setter: GetCapturedCard(), SetCapturedCard()
  - Add parameter 'entityID' tofunction 'AddEnchantment()': The entity ID to be stored in the enchantment.
  - Replace variable 'id' in class 'Entity' with 'GameTag::ENTITY_ID'
  - Add variable 'm_useEntityID' to class 'AddEnchantmentTask': The flag that indicates whether it uses entity ID.
  - Add variable 'm_topdeckTask' to class 'Power'
    - Related methods
      - GetTopdeckTask(): Returns a list of topdeck tasks.
      - AddTopdeckTask(): Adds topdeck task.
  - Add code to process 'EntityType::DECK' to class 'AddCardTask'
  - Change return type of function 'AddCardToHand()' to 'bool': The flag that indicates a card is successfully added to hand.
  - Add variable 'm_progressType' to class 'QuestProgressTask'
    - Add enum 'ProgressType': The type of quest progress.
  - Add some enum values
    - 'TargetingType::CHARACTERS_EXCEPT_HERO'
    - 'CharacterType::CHARACTERS_EXCEPT_HERO'
    - 'GameTag::CAN_TARGET_MINION_BY_HERO_POWER'
  - Change code to set game tags from minion's instead of card's
  - Add code to reset choice and add entity to setaside zone

### CODE REVIEW

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/Rosetta/Tasks/SimpleTasks/ActivateCapturedDeathrattleTask.cpp  1
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  1372
- Sources/Rosetta/Tasks/SimpleTasks/DiscoverTask.cpp  1
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  19
- Sources/Rosetta/CardSets/CoreCardsGen.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5127637&bid=16852975)

File: Sources/Rosetta/CardSets/DragonsCardsGen.cpp [hunk 1]
```diff
@@ -577,11 +680,35 @@ void DragonsCardsGen::AddDruidNonCollect(std::map<std::string, CardDef>& cards)
...
+    power.AddPowerTask(
+        std::make_shared<AddEnchantmentTask>("DRG_311e", EntityType::TARGET));
+    cards.emplace(
+        "DRG_311b",
+        CardDef(power, PlayReqs{ { PlayReq::REQ_MINION_TARGET, 0 },
+                                 { PlayReq::REQ_TARGET_TO_PLAY, 0 } }));
+
+    // ------------------------------------------ SPELL - DRUID
+    // [DRG_311e] Spore Hardened (*) - COST:0
+    // - Set: Expert1
```
Comment (ShyRoute) [inline Sources/Rosetta/CardSets/DragonsCardsGen.cpp]: Set을 Dragons로 수정해야 할 것 같습니다.

File: Sources/Rosetta/CardSets/DragonsCardsGen.cpp [hunk 2]
```diff
@@ -1349,6 +1713,11 @@ void DragonsCardsGen::AddRogue(std::map<std::string, CardDef>& cards)
     // - DISCOVER = 1
     // - USE_DISCOVER_VISUALS = 1
     // --------------------------------------------------------
+    power.ClearData();
+    power.AddPowerTask(
+        std::make_shared<DiscoverTask>(CardType::MINION, CardClass::INVALID,
+                                       Race::INVALID, Rarity::LEGENDARY));
```
Comment (ShyRoute) [inline Sources/Rosetta/CardSets/DragonsCardsGen.cpp]: 도적(Rogue) 외 다른 직업의 전설 카드를 발견해야 하는데 CardClass::INVALID로 되어있습니다.

File: Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp [hunk 3]
```diff
@@ -1316,93 +1392,2618 @@ TEST_CASE("[Paladin : Spell] - DRG_008 : Righteous Cause")
...
+    CHECK_EQ(curField[1]->GetHealth(), 1);
+    CHECK_EQ(curField[2]->GetHealth(), 6);
+
+    game.Process(curPlayer, EndTurnTask());
+    game.ProcessUntil(Step::MAIN_START);
+
+    game.Process(opPlayer, PlayCardTask::Minion(card4));
+    CHECK_EQ(curField[0]->GetHealth(), 1);
+    CHECK_EQ(curField[1]->GetHealth(), 1);
+    CHECK_EQ(curField[2]->GetHealth(), 1);
```
Comment (ShyRoute) [inline Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: Venarus를 낸 플레이어(opPlayer) 의 하수인은 체력이 1로 변경되지 않는 상황에 대한 테스트가 필요하지 않을까요?

File: Sources/Rosetta/CardSets/DragonsCardsGen.cpp [hunk 4]
```diff
@@ -714,13 +857,72 @@ void DragonsCardsGen::AddHunter(std::map<std::string, CardDef>& cards)
...
+                    if (player == player->game->GetCurrentPlayer())
+                    {
+                        pick->SetExhausted(true);
+                    }
+
+                    break;
+                }
+
+                return std::vector<Playable*>{};
+            })
```
Comment (1cekrim) [inline Sources/Rosetta/CardSets/DragonsCardsGen.cpp]: 지금은 playables에서 뽑은 카드가  SecretZone에 이미 있다면 그걸 playables에서 제거하고, 다시 한 장을 뽑는 걸 반복하는 구조로 되어있는데,
playables에서 SecretZone에 있는 비밀을 전부 제거한 다음, 한 장을 뽑는 방식으로 로직을 바꾸면 코드가 더 간결해 질 것 같습니다.

```
        std::make_shared<FuncPlayableTask>(
            [=](const std::vector<Playable*>& playables) {
                if (playables.empty())
                {
                    return std::vector<Playable*>{};
                }

                auto player = playables[0]->player;

                EraseIf(const_cast<std::vector<Playable*>&>(playables),
                        [=](Playable* playable) {
                            return player->GetSecretZone()->Exist(playable);
                        });

                if (playables.empty())
                {
                    return std::vector<Playable*>{};
                }

                auto pick = *Random::get(playables);

                // Remove it from deck zone
                player->GetDeckZone()->Remove(pick);
                if (auto trigger = pick->card->power.GetTrigger(); trigger)
                {
                    trigger->Activate(pick);
                }

                // Add it to secret zone
                player->GetSecretZone()->Add(pick);
                if (player == player->game->GetCurrentPlayer())
                {
                    pick->SetExhausted(true);
                }

                return std::vector<Playable*>{};
            })
```

File: Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp [hunk 5]
```diff
@@ -1316,93 +1392,2618 @@ TEST_CASE("[Paladin : Spell] - DRG_008 : Righteous Cause")
...
+
+    game.Process(opPlayer, PlayCardTask::SpellTarget(card4, card1));
+    bool check1 = (minion2->GetAttack() == 7 && minion2->GetHealth() == 9 &&
+                   minion2->HasDeathrattle() == true &&
+                   minion3->GetAttack() == 4 && minion3->GetHealth() == 6);
+    bool check2 = (minion2->GetAttack() == 4 && minion2->GetHealth() == 6 &&
+                   minion3->GetAttack() == 7 && minion3->GetHealth() == 9 &&
+                   minion3->HasDeathrattle() == true);
+    bool check = check1 || check2;
+    CHECK_EQ(check, true);
```
Comment (ShyRoute) [inline Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: `CHECK_EQ(check1 || check2, true);` 로 줄여도 될 것 같습니다.

File: Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp [hunk 6]
```diff
@@ -1316,93 +1392,2618 @@ TEST_CASE("[Paladin : Spell] - DRG_008 : Righteous Cause")
...
+
+    Player* curPlayer = game.GetCurrentPlayer();
+    Player* opPlayer = game.GetOpponentPlayer();
+    curPlayer->SetTotalMana(10);
+    curPlayer->SetUsedMana(0);
+    opPlayer->SetTotalMana(10);
+    opPlayer->SetUsedMana(0);
+
+    const auto card1 =
+        Generic::DrawCard(curPlayer, Cards::FindCardByName("Dragon's Hoard")
```
Comment (ShyRoute) [inline Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: 카드들의 Class가 도적이 아닌지에 대한 테스트가 필요할 것 같습니다.

Comment (ShyRoute) [inline Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: `|` 를 `||` 로 고쳐주세요.

Comment (ShyRoute) [inline Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: check4 부분 `|`를 `||`로 수정해야 할 것 같습니다.

Comment (ShyRoute) [reply Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: 코드 위치가 이상하게 나오네요. 파일 기준 3813 ~ 3821줄 부분입니다.

Comment (ArtBlnd) [reply Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: 이건 다른 이야기인데 아마 이 코드에서는 | 로 하나 ||로 하나 차이가 없을거에요.

File: Sources/Rosetta/CardSets/DragonsCardsGen.cpp (hunk 1 above)
Comment (utilForever) [reply Sources/Rosetta/CardSets/DragonsCardsGen.cpp]: 수정했습니다.

File: Sources/Rosetta/CardSets/DragonsCardsGen.cpp (hunk 2 above)
Comment (utilForever) [reply Sources/Rosetta/CardSets/DragonsCardsGen.cpp]: 수정했습니다.

File: Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp (hunk 3 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: 테스트 추가했습니다.

File: Sources/Rosetta/CardSets/DragonsCardsGen.cpp (hunk 4 above)
Comment (utilForever) [reply Sources/Rosetta/CardSets/DragonsCardsGen.cpp]: 수정했습니다.

File: Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp (hunk 6 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: 테스트 추가했습니다.

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  1365
- Sources/Rosetta/CardSets/CoreCardsGen.cpp  2
- Sources/Rosetta/Tasks/SimpleTasks/ActivateCapturedDeathrattleTask.cpp  1
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  19
- Sources/Rosetta/Tasks/SimpleTasks/DiscoverTask.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5127637&bid=16852975)

File: Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp (hunk 6 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: 잘못된 코드가 아니라서 패스합니다.

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/Rosetta/Tasks/SimpleTasks/DiscoverTask.cpp  1
- Sources/Rosetta/Tasks/SimpleTasks/ActivateCapturedDeathrattleTask.cpp  1
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  19
- Sources/Rosetta/CardSets/CoreCardsGen.cpp  2
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  1368
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5127637&bid=16852975)

**Pull request merged**